include/ghex/transport_layer/raw_shared_message_buffer.hpp

```diff
@@ -0,0 +1,221 @@
+/* 
+ * GridTools
+ * 
+ * Copyright (c) 2014-2019, ETH Zurich
+ * All rights reserved.
+ * 
+ * Please, refer to the LICENSE file in the root directory.
+ * SPDX-License-Identifier: BSD-3-Clause
+ * 
+ */
+#ifndef INCLUDED_GHEX_TL_SHARED_MESSAGE_BUFFER_HPP
+#define INCLUDED_GHEX_TL_SHARED_MESSAGE_BUFFER_HPP
+
+#include <memory>
+#include "./message_buffer.hpp"
+
+namespace gridtools {
+    namespace ghex {
+        namespace tl {
+
+	    /* A struct that keeps the message with the refcount.
+	       The shared message buffer keeps a pointer to a dynamically
+	       allocated refcounted_message. That pointer can be shared
+	       between many messages. 
+	     */
```

Comment:
This has implications on thread safety, right? Either this or next comment should clarify. We have many pieces around and it's difficult for me to keep all them in context. We should prepare a presentation for the others in order to put everything in perspective (I would need it to glue together stuff)